File: solutions/cpp/collatz-conjecture/collatz_conjecture.cpp
#include "collatz_conjecture.h"

#include <stdexcept>

namespace collatz_conjecture {

    auto steps(int n) -> int
    {
        if (n <= 0)
            throw std::domain_error("");

        int steps = 0;

        while (n != 1)
        {
            if (n % 2)
                n = 3 * n + 1;
            else
                n /= 2;

            steps++;
        }

        return steps;
    }

}  // namespace collatz_conjecture

File: solutions/csharp/collatz-conjecture/CollatzConjecture.cs
public static class CollatzConjecture
{
    public static int Steps(int n)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(n);

        int s = 0;

        while (n > 1)
        {
            n = n % 2 == 0 ? n / 2 : 3 * n + 1;
            s += 1;
        }

        return s;
    }
}